mlir-sycl/test/Transforms/sycl-method-to-sycl-call.mlir

```diff
@@ -8,35 +8,73 @@
+// CHECK-NEXT:      %[[VAL_5:.*]] = llvm.addrspacecast %[[VAL_4]] : !llvm.ptr<!sycl_accessor_2_i32_rw_gb> to !llvm.ptr<!sycl_accessor_2_i32_rw_gb, 4>
+// CHECK-NEXT:      %[[VAL_6:.*]] = "polygeist.pointer2memref"(%[[VAL_5]]) : (!llvm.ptr<!sycl_accessor_2_i32_rw_gb, 4>) -> memref<?x!sycl_accessor_2_i32_rw_gb, 4>
+// CHECK-NEXT:      %[[VAL_7:.*]] = sycl.call(%[[VAL_6]], %[[VAL_1]]) {FunctionName = @"operator[]", MangledFunctionName = @_ZNK4sycl3_V18accessorIiLi2ELNS0_6access4modeE1026ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi2EvEERiNS0_2idILi2EEE, TypeName = @accessor} : (memref<?x!sycl_accessor_2_i32_rw_gb, 4>, memref<?x!sycl_id_2_>) -> memref<?xi32, 4>
+// CHECK-NEXT:      return %[[VAL_7]] : memref<?xi32, 4>
+// CHECK-NEXT:    }
 
-func.func @accessor_subscript_operator(%arg0: memref<?x!sycl_accessor_2_i32_rw_gb, 4>, %arg1: memref<?x!sycl_id_2_>) -> memref<?xi32, 4> {
-  %0 = sycl.accessor.subscript %arg0[%arg1] {BaseType = memref<?x!sycl_accessor_2_i32_rw_gb, 4>, FunctionName = @"operator[]", MangledFunctionName = @_ZNK4sycl3_V18accessorIiLi2ELNS0_6access4modeE1026ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi2EvEERiNS0_2idILi2EEE, TypeName = @accessor} : (memref<?x!sycl_accessor_2_i32_rw_gb, 4>, memref<?x!sycl_id_2_>) -> memref<?xi32, 4>
+func.func @accessor_subscript_operator(%arg0: !sycl_accessor_2_i32_rw_gb, %arg1: memref<?x!sycl_id_2_>) -> memref<?xi32, 4> {
+  %0 = sycl.accessor.subscript %arg0[%arg1] {BaseType = memref<?x!sycl_accessor_2_i32_rw_gb, 4>, FunctionName = @"operator[]", MangledFunctionName = @_ZNK4sycl3_V18accessorIiLi2ELNS0_6access4modeE1026ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi2EvEERiNS0_2idILi2EEE, TypeName = @accessor} : (!sycl_accessor_2_i32_rw_gb, memref<?x!sycl_id_2_>) -> memref<?xi32, 4>
   return %0 : memref<?xi32, 4>
 }
 
-// CHECK-LABEL: func.func @range_get(%arg0: memref<?x!sycl_range_2_, 4>, %arg1: i32) -> i64 {
-// CHECK-NEXT: %0 = sycl.cast(%arg0) : (memref<?x!sycl_range_2_, 4>) -> memref<?x!sycl_array_2_, 4>
-// CHECK-NEXT: %1 = sycl.call(%0, %arg1) {FunctionName = @get, MangledFunctionName = @_ZNK4sycl3_V16detail5arrayILi2EE3getEi, TypeName = @array} : (memref<?x!sycl_array_2_, 4>, i32) -> i64
+// CHECK-LABEL:   func.func @range_get(
+// CHECK-SAME:                         %[[VAL_0:.*]]: !sycl_range_2_,
+// CHECK-SAME:                         %[[VAL_1:.*]]: i32) -> i64 {
+// CHECK-NEXT:      %[[VAL_2:.*]] = arith.constant 0 : index
+// CHECK-NEXT:      %[[VAL_3:.*]] = memref.alloca() : memref<1x!sycl_range_2_>
+// CHECK-NEXT:      memref.store %[[VAL_0]], %[[VAL_3]]{{\[}}%[[VAL_2]]] : memref<1x!sycl_range_2_>
+// CHECK-NEXT:      %[[VAL_4:.*]] = "polygeist.memref2pointer"(%[[VAL_3]]) : (memref<1x!sycl_range_2_>) -> !llvm.ptr<!sycl_range_2_>
```

Comment:
I thought you are doing 
```
%0 = memref.alloca 
%1 = memref.cast(%0) -> memref<?x...>
polygeist.memref2pointer(%1) : (memref<?x...>) -> !llvm.ptr<...?
```
why is it safe to fold and why is there a sycl.cast : (memref<?x...>) -> memref<?x...> at line 37.

Reply:
We're casting to the base class. `sycl.cast` cannot have same src and dst types, right?

I removed all folds but the one when creating the constant: they won't be performed anyway.

Reply:
Given:
```
"polygeist.memref2pointer"(%[[VAL_3]]) : (memref<1x!sycl_range_2_>) -> !llvm.ptr<!sycl_range_2_>
```
this operation is changing a memref to a pointer, however I think this operation ought to be legal only when the memref has the pointer form, that is  `memref<?x!sycl_range_2_>` so it should be proceeded by a memref.cast in order to case  `memref<1x!sycl_range_2_>` -> `memref<?x!sycl_range_2_>`

At least that is the convention we have been using all along.

Reply:
Right, that seems to be what the code does, wonder if he hasn't updated the tests yet.

Reply:
It is being inserted, but it is being canonicalized-out (weird, as we're using `create` for every operation). I think I'll use `createOrFold` for the cast anyway, as we might find that the shapes do not differ and we can safe a cast that way.

I was not aware of this convention, but it is not being followed in lots of places in the code (just compile the `functions.cpp` example with MLIR output and see many examples there).

Reply:
The problem is that is unknown what the exact semantics of `polygeist.memref2pointer` are. We know that in general a memref type is not equivalent to an llvm pointer so, I think it would be wise to restrict the type of memrefs that can be legally converted to LLVM pointers via the use of `polygeist.memref2pointer` by letting only memref<?xelem_type> be allowed to be converted to a pointer.  Now... there are places where the convention is not followed. I think we should fix those cases (not in this PR of course).

Reply:
I totally agree with you: that operation is not clear at all (I'd even say its `pointer2memref` counterpart violates the very logic of `memrefs`). We should discuss all of this offline.